### ISSUE TITLE
Performance discussion

### ISSUE THREAD

**Issue by lexruster** [lppkarl/Trady#37]
Hi, @lppkarl 
After first research of library i started some more serious development :)
And first and the biggest problem that I faced now is performance.
i just run some code in release configuration
``` C#
             var buyRule = Rule.Create(c => c.IsEmaBullishCross(3, 7));
            var sellRule = Rule.Create(c => c.IsEmaBearishCross(3, 7));

            var importer = new CsvImporter(setting.CsvPath);
            IReadOnlyList<IOhlcv> candles = await importer.ImportAsync(setting.Symbol);
            _logger.LogInformation($"Readed {candles.Count} of record, start backtest: ", setting);

            var runner = new Builder()
                .Add(candles)
                .Buy(buyRule )
                .Sell(sellRule )
                .Build();

            Result result = await runner.RunAsync(setting.Amount);
```
Where i have just 1 month of candles for each minute, it is 40 320 candles.
And it is executed 15 minutes. Looks like all time was consumed by evaluating rules.
Is that something that you expect? Did you plan to use the library for those volumes of data?

I also playing with https://enigmampc.github.io/catalyst/index.html
but i  don't like Python :) Will comment if i will test the same calculation with Catalyst.

Do you have any thoughts or plans about speedup?

**Comment by irperez** [comment]
@lexruster can you clone the repo and run your test with a profiler?  Can you submit some screenshots of where the hotspots are at?  Or at the very least submit your sample data to a public folder where I can run your sample locally and check for myself.  Possibly in csv format.

**Comment by BergWerkGIS** [comment]
@lppkarl @lexruster I'm hitting a bottleneck when calculating `MacdLine` with `Trady@3.0.0`.

The profiler screenshots are from a session with 135 datasets each containing 720 candles.
Almost ~50% of CPU time are spent calculating `MacdLine`.
No difference between `Debug` and `Release`.

![image](https://user-images.githubusercontent.com/4549888/39425663-c3c32456-4c7c-11e8-9cac-c6bed6e0a015.png)
![image](https://user-images.githubusercontent.com/4549888/39425666-c8014796-4c7c-11e8-8361-38e895c600ed.png)

Does this help in pinpointing the problem?

I could provide a sample project but it would take quite some effort and time on my side to extract needed logic and data from my (pet) project (quite complex already, custom data storage, ...).

**Comment by BergWerkGIS** [comment]
It's not `MacdLine` specifically but any `Macd` indicator.

If I change the order of calling `MacdLine`, `SignalLine` and `MacdHistogram` it's always the first one that takes most of the time:

![image](https://user-images.githubusercontent.com/4549888/39426187-260b72ce-4c7f-11e8-8b62-9f9d5cd4bf81.png)


![image](https://user-images.githubusercontent.com/4549888/39426224-46d9cbb8-4c7f-11e8-9081-d0ea73b840b5.png)

**Comment by irperez** [comment]
I'm wondering if this is related to issue #60 

In that scenario, LINQ is the root cause in that case.

**Comment by lppkarl** [comment]
@irperez yeah, i was also wondering if LINQ has been the major issue behind this case, but i need further check on this

**Comment by lppkarl** [comment]
@irperez I've just added ToList() in Ema calculation which relies by Macd indicator, the performance boost is very significant from 156697ms to 631ms (37767 candles, 248x boost), the lazy-evaluation of LINQ seems to be the major cause of the great performance hit. I might have to scan through the indicators and see if there's room for performance optimization regarding to the lazy-evaluation problem.

**Comment by irperez** [comment]
@ippkarl it might be worth while to add some benchmarks on the indicators. I can look into this using benchmark.net.

**Comment by lppkarl** [comment]
@irperez Okay, thanks for the help. 👍

**Comment by irperez** [comment]
@lppkarl I just added a pull request (#62 ) with some benchmarks.  Looks like Dymoi maybe another problem.

**Comment by lppkarl** [comment]
@irperez thank you very much for the benchmark, seems dymoi really takes too much time to compute when compared to the others, i'll take a look of it first.
